src/main/java/io/jenkins/plugins/gitlabbranchsource/servers/GitLabServer.java

```diff
@@ -0,0 +1,289 @@
+     * of hooks.
+     *
+     * @return the {@link StandardUsernamePasswordCredentials#getId()} of the credentials to use for auto-management
+     * of hooks.
+     */
+    @CheckForNull
+    public String getCredentialsId() {
+        return credentialsId;
+    }
+
+    /**
+     * Constructor
+     *
+     * @param displayName   Optional name to use to describe the end-point.
+     * @param serverUrl     The URL of this GitLab Server
+     * @param manageHooks   {@code true} if and only if Jenkins is supposed to auto-manage hooks for this end-point.
+     * @param credentialsId The {@link StandardUsernamePasswordCredentials#getId()} of the credentials to use for
+     *                      auto-management of hooks.
+     * @since 1.0.5
+     */
+    @DataBoundConstructor
+    public GitLabServer(@CheckForNull String displayName, @NonNull String serverUrl, boolean manageHooks,
+                       @CheckForNull String credentialsId) {
+        this.manageHooks = manageHooks;
+        this.credentialsId = credentialsId;
+        this.serverUrl = defaultIfBlank(serverUrl, GITLAB_SERVER_URL);
+        this.name = StringUtils.isBlank(displayName)
+                ? SCMName.fromUrl(this.serverUrl, COMMON_PREFIX_HOSTNAMES)
+                : displayName;
+    }
+
+
+    /**
+     * Looks up the {@link StandardCredentials} to use for auto-management of hooks.
+     *
+     * @return the credentials or {@code null}.
+     */
+    @CheckForNull
+    public StandardCredentials credentials() {
+        return StringUtils.isBlank(credentialsId) ? null : CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(
+                        StandardCredentials.class,
+                        Jenkins.getActiveInstance(),
+                        ACL.SYSTEM,
+                        fromUri(serverUrl).build()
+                ),
+                CredentialsMatchers.allOf(
+                        AuthenticationTokens.matcher(GitLabAuth.class),
+                        withId(credentialsId)
+                )
+        );
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public DescriptorImpl getDescriptor() {
+        return (DescriptorImpl) super.getDescriptor();
+    }
+
+
+    /**
+     * Our descriptor.
+     */
+    @Extension
+    public static class DescriptorImpl extends Descriptor<GitLabServer> {
+
+        @NonNull
+        @Override
+        public String getDisplayName() {
+            return "GitLab Server";
+        }
+
+
+        public String getAdvanceConfigMessage() {
+            return Messages.GitLabServer_advancedSectionForFuture();
+        }
+
+        /**
+         * Checks that the supplied URL is valid.
+         *
+         * @param value the URL to check.
+         * @return the validation results.
+         */
+        public static FormValidation doCheckServerUrl(@QueryParameter String value) {
+            // TODO: Add support for GitLab Ultimate (self hosted) and Gold (saas)
```

Comment:
Do you have Jira stories or tasks representing these? It's good practice to create them for significant TODO's to make sure they don't get forgotten. You could even include the IDs here for reference

Reply:
Yes I have stories for them, I will add the reference.